File: docs-examples/example-java/src/test/java/io/micronaut/jms/docs/exceptions/ErrorThrowingConsumer.java
package io.micronaut.jms.docs.exceptions;

import io.micronaut.context.annotation.Requires;

// tag::imports[]
import io.micronaut.jms.annotations.JMSListener;
import io.micronaut.jms.annotations.Queue;
import io.micronaut.jms.listener.JMSListenerErrorHandler;
import io.micronaut.messaging.annotation.MessageBody;

import javax.inject.Singleton;
import javax.jms.Message;
import javax.jms.Session;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;

import static io.micronaut.jms.activemq.classic.configuration.ActiveMqClassicConfiguration.CONNECTION_FACTORY_BEAN_NAME;
// end::imports[]

@Requires(property = "spec.name", value = "ErrorHandlingSpec")
// tag::clazz[]
@JMSListener(value = CONNECTION_FACTORY_BEAN_NAME, errorHandlers = {AccumulatingErrorHandler.class, CountingErrorHandler.class}) // <1>
class ErrorThrowingConsumer {

    Collection<String> messages = Collections.synchronizedSet(new HashSet<>());

    @Queue(value = "error-queue", concurrency = "1-1", errorHandlers = {CountingErrorHandler.class}) // <2>
    void receive(@MessageBody String message) {
        if ("throw an error".equalsIgnoreCase(message)) {
            throw new RuntimeException("this is an error"); // <3>
        }
        messages.add(message);
    }
}
// end::clazz[]

@Requires(property = "spec.name", value = "ErrorHandlingSpec")
@Singleton
class CountingErrorHandler implements JMSListenerErrorHandler {

    Integer count = 0;

    public void handle(Session session, Message message, Throwable ex) {
        count++;
    }

}

@Requires(property = "spec.name", value = "ErrorHandlingSpec")
@Singleton
class AccumulatingErrorHandler implements JMSListenerErrorHandler {

    Collection<Throwable> exceptions = Collections.synchronizedList(new ArrayList<>());

    public void handle(Session session, Message message, Throwable ex) {
        if (ex != null) {
            exceptions.add(ex);
        }
    }

}

File: docs-examples/example-python/src/test/python/micronaut/jms/docs/exceptions/ErrorThrowingConsumer.py
# tag::imports[]
from typing import Annotated

from jakarta.inject import Singleton
from jakarta.jms import Message, Session
from java.lang import RuntimeException, Throwable
from micronaut.context.annotation import Requires
from micronaut.jms.annotations import JMSListener, Queue
from micronaut.jms.listener import JMSListenerErrorHandler
from micronaut.messaging.annotation import MessageBody
# end::imports[]


@Requires(property="spec.name", value="ErrorHandlingSpec")
@Singleton
class CountingErrorHandler(JMSListenerErrorHandler):

    def __init__(self) -> None:
        self.count = 0

    def handle(self, session: Session, message: Message, ex: Throwable) -> None:
        self.count += 1


@Requires(property="spec.name", value="ErrorHandlingSpec")
@Singleton
class AccumulatingErrorHandler(JMSListenerErrorHandler):

    def __init__(self) -> None:
        self.exceptions: list[Throwable] = []

    def handle(self, session: Session, message: Message, ex: Throwable) -> None:
        if ex is not None:
            self.exceptions.append(ex)


@Requires(property="spec.name", value="ErrorHandlingSpec")
# tag::clazz[]
@JMSListener("activeMqConnectionFactory", errorHandlers=[AccumulatingErrorHandler])  # <1>
class ErrorThrowingConsumer:

    def __init__(self) -> None:
        self.messages: set[str] = set()

    @Queue("error-queue", errorHandlers=[CountingErrorHandler])  # <2>
    def receive(self, message: Annotated[str, MessageBody]) -> None:
        if message.lower() == "throw an error":
            raise RuntimeException("this is an error")  # <3>
        self.messages.add(message)
# end::clazz[]
